libyul/backends/wasm/polyfill/eei.h

```diff
@@ -0,0 +1,524 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * This file includes logical functions for the wasm polyfill.
+ */
+
+#pragma once
+
+#include <string>
+
+namespace solidity::yul::wasm::polyfill
+{
+
+static std::string const eei{R"(
+// solidity::yul::wasm::polyfill::eei
+
+// address of the current contract / execution context
+function address() -> z1, z2, z3, z4 {
+	// eth.getAddress: Gets address of currently executing account and stores it
+	// in memory at the given offset. This address is encoded as a 160 bit number,
+	// represented as a 20 bytes long little endian unsigned integer in memory.
+	eth.getAddress(4:i32)
+
+	// mload_address takes the little endian unsigned integer from memory and converts
+	// it to an big endian unsigned integer (evm is big endian).
+	z2, z3, z4 := mload_address(0:i32)
+}
+
+//  balance(a): wei balance at address a
+function balance(x1, x2, x3, x4) -> z1, z2, z3, z4 {
+	// address x is a big endian encoded unsigned integer.
+	// mstore_address(pos, address) will write that address as a little endian
+	// unsigned integer at offset 0.
+	mstore_address(0:i32, x1, x2, x3, x4)
+
+	// getExternalBalance(addressOffset, resultOffset): Gets balance of the
+	// given account and loads it into memory at the given offset.
+	// address need to be encoded as a little endian unsigned integer.
+	// balance is of type u128 represented as a 16 bytes long
+	// little endian unsigned integer in memory.
+	eth.getExternalBalance(12:i32, 32:i32)
+
+	z3 := i64.load(40:i32)
```

Comment:
I don't really understand why this actually works. 

If `z1, z2, z3, z4` are `big endian`, where the data is currently encoded as little endian, should'nt this be `z3 := bswap64(i64.load(40))` and `z4 := bswap64(i64.load(32))` - or, if that data is already stored as `big endian`, shouldn't this be `z3 := i64.load(32)` and `z4 := i64.load(40)`. 

Both variants are not working. 
 
I don't understand why the `i64`'s need to get loaded "reversely" without the need of using `bswap64`.